### PR TITLE
[SRVCOM-1237] Remove redundant service monitors during upgrades

### PR DESCRIPTION
**Pull request by skonto**
- fixes issue faced during upgrades
- tested it also via creating a sm "knative-openshift-metrics-2" and restarting the knative-openshift pod

### CODE REVIEW

File: knative-operator/pkg/common/monitoring.go [hunk 1]
```diff
@@ -36,6 +42,36 @@ func SetupMonitoringRequirements(api client.Client, instance mf.Owner) error {
 	return nil
 }
 
+func RemoveOldServiceMonitorResources(namespace string, api client.Client) error {
+	currentSMName := os.Getenv(operatorServiceMonitorNameEnvKey)
+	if currentSMName == "" {
+		currentSMName = operatorServiceMonitorDefaultName
+	}
```
Comment (markusthoemmes) [inline knative-operator/pkg/common/monitoring.go:49]: Is the defaulting necessary? Seems like we use explicit values everywhere, so I'd rather drop this to catch accidental issues.

File: knative-operator/pkg/common/monitoring.go [hunk 2]
```diff
@@ -36,6 +42,36 @@ func SetupMonitoringRequirements(api client.Client, instance mf.Owner) error {
...
+	smList := monitoringv1.ServiceMonitorList{}
+	if err := api.List(context.TODO(), &smList, client.InNamespace(namespace)); err != nil {
+		return err
+	}
+	for _, sm := range smList.Items {
+		// Skip the sm that is being installed with the current operator version
+		if sm.Name != currentSMName && strings.HasPrefix(sm.Name, "knative-openshift-metrics-") {
+			// Delete service monitor and the related service monitor service, skip error if not found
+			if err := api.Delete(context.TODO(), sm); err != nil && !errors.IsNotFound(err) {
```
Comment (markusthoemmes) [inline knative-operator/pkg/common/monitoring.go:58]:
```suggestion
			if err := api.Delete(context.Background(), sm); err != nil && !errors.IsNotFound(err) {
```

File: knative-operator/pkg/common/monitoring.go [hunk 3]
```diff
@@ -36,6 +42,36 @@ func SetupMonitoringRequirements(api client.Client, instance mf.Owner) error {
...
+				return err
+			}
+			service := v1.Service{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: namespace,
+					Name:      sm.Name,
+				},
+			}
+			if err := api.Delete(context.TODO(), &service); err != nil && !errors.IsNotFound(err) {
```
Comment (markusthoemmes) [inline knative-operator/pkg/common/monitoring.go:67]:
```suggestion
			if err := api.Delete(context.Background(), &service); err != nil && !errors.IsNotFound(err) {
```

File: knative-operator/pkg/common/monitoring.go [hunk 4]
```diff
@@ -36,6 +42,36 @@ func SetupMonitoringRequirements(api client.Client, instance mf.Owner) error {
 	return nil
 }
 
+func RemoveOldServiceMonitorResources(namespace string, api client.Client) error {
+	currentSMName := os.Getenv(operatorServiceMonitorNameEnvKey)
+	if currentSMName == "" {
+		currentSMName = operatorServiceMonitorDefaultName
+	}
+	smList := monitoringv1.ServiceMonitorList{}
+	if err := api.List(context.TODO(), &smList, client.InNamespace(namespace)); err != nil {
```
Comment (markusthoemmes) [inline knative-operator/pkg/common/monitoring.go:51]: I sort of wondered if it'd be simpler to enumerate the items we want to delete explicitly, but not a strong opinion, this approach looks viable as well.

**Review by markusthoemmes — COMMENTED** [review]
Do we know if this is "only" an issue with our very first Service/SM and the subsequent ones (managed by OLM) are fine being handled by OLM?

If so, I'd target this towards deleting just this very first set of resources. If OLM doesn't garbage collect the newer stuff, this approach is fine too.

File: knative-operator/pkg/common/monitoring.go (hunk 4 above)
Comment (skonto) [reply knative-operator/pkg/common/monitoring.go:51]: I kept this a bit generic not having to define them per release.

File: knative-operator/pkg/common/monitoring.go (hunk 1 above)
Comment (skonto) [reply knative-operator/pkg/common/monitoring.go:49]: I could remove it yes.

**Comment by skonto** [comment]
> Do we know if this is "only" an issue with our very first Service/SM and the subsequent ones (managed by OLM) are fine being handled by OLM?

I will double check and adapt the PR if we only need to remove the very first one.

File: knative-operator/pkg/common/monitoring.go (hunk 2 above)
Comment (skonto) [reply knative-operator/pkg/common/monitoring.go:58]: I was not sure if I could pass any ctx later on thus TODO but Background probably makes more sense.

Comment (markusthoemmes) [reply knative-operator/pkg/common/monitoring.go:58]: Yeah the line is somewhat blurry, but TODO shouldn't be in any production code (IMO at least). It's supposed to trigger linter warnings I guess :shrug:

**Review by markusthoemmes — APPROVED** [review]
/lgtm
/approve

**Pull request merged**